### ISSUE TITLE
ChainExpression

### ISSUE THREAD

**Issue by neumatho** [meriyah/meriyah#98]
Hi,

For the expression **(a?.import("string")?.import.meta??(a))**, the argument "string" is stored as a ChainExpression node with a Literal in expression property. The ChainExpression should be skipped and the Literal should be stored directly in arguments instead. According to the AST specifications, it is not allowed to have a Literal in expression property in a ChainExpression node:

https://github.com/estree/estree/blob/master/es2020.md#chainexpression

-Thomas

**Comment by KFlash** [comment]
Cc @3cp

**Comment by neumatho** [comment]
The same problem with **a?.[x]**. Here you create an _ExpressionStatement_ with a _MemberExpression_ in _expression_ and a _ChainExpression_ in _property_. It should be the other way around. _ExpressionStatement_ with a _ChainExpression_ in _expression_ and the _MemberExpression_ in _expression_.

**Comment by KFlash** [comment]
Sounds reasonable. As said before some students of mine did this code, and I personally avoid ESTree. 

@aladdin-add Ideas?

**Comment by KFlash** [comment]
@neumatho So I get you correct. The result you want is like [this](https://escaya.github.io/escaya/?code=a%3F.%5Bx%5D&method=parse&range=undefined&loc=false&next=false&cst=false&module=false&raw=undefined&jsx=undefined&directives=undefined&attachComment=undefined&webCompat=undefined&lexical=undefined) but with ESTree AST ?

**Comment by neumatho** [comment]
I want it like this, which is the way Acorn generates it and seems to comply with the specification:

{
  "type": "Program",
  "body": [
    {
      "type": "ExpressionStatement",
      "expression": {
        "type": "ChainExpression",
        "expression": {
          "type": "MemberExpression",
          "object": {
            "type": "Identifier",
            "name": "a"
          },
          "property": {
            "type": "Identifier",
            "name": "x"
          },
          "computed": true,
          "optional": true
        }
      }
    }
  ],
  "sourceType": "module"
}

**Comment by neumatho** [comment]
and for (a?.import("string")?.import.meta??(a)) is should look like this:

```
{
  "type": "Program",
  "body": [
    {
      "type": "ExpressionStatement",
      "expression": {
        "type": "LogicalExpression",
        "left": {
          "type": "ChainExpression",
          "expression": {
            "type": "MemberExpression",
            "object": {
              "type": "MemberExpression",
              "object": {
                "type": "CallExpression",
                "callee": {
                  "type": "MemberExpression",
                  "object": {
                    "type": "Identifier",
                    "name": "a"
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "import"
                  },
                  "computed": false,
                  "optional": true
                },
                "arguments": [
                  {
                    "type": "Literal",
                    "value": "string",
                    "raw": "\"string\""
                  }
                ],
                "optional": false
              },
              "property": {
                "type": "Identifier",
                "name": "import"
              },
              "computed": false,
              "optional": true
            },
            "property": {
              "type": "Identifier",
              "name": "meta"
            },
            "computed": false,
            "optional": false
          }
        },
        "operator": "??",
        "right": {
          "type": "Identifier",
          "name": "a"
        }
      }
    }
  ],
  "sourceType": "module"
}
```

**Comment by 3cp** [comment]
I did the migration from old estree spec to the new spec ChainExpression.
I can reproduce all the bugs here, clearly we missed some test coverage. Will need some time to understand where went wrong, hopefully not as nasty as those known bugs in destruction.

**Comment by 3cp** [comment]
The bug is from I used a global flag here.
https://github.com/meriyah/meriyah/blob/d3de211ef404cc10b736e853507e2f93491e4d4f/src/parser.ts#L3964

This is to record whether an expression contains optional chain, so that later we can wrap the whole expression in ChainExpression node according to the spec.

But the usage of global flag is wrong, in `a?.[x]`, there are actually two expressions: `a?.[]` and a sub expression `x`. The sub expression needs to be parsed independently, but now the parser saw the global flag (set by previous `?.` token) and mistakenly think `x` needs wrap.

@KFlash any suggestion on how to parse it correctly? Some flag to be saved with the expression itself, rather than globally?

**Comment by KFlash** [comment]
Maybe try to look at the [Escaya source](https://github.com/escaya/escaya/blob/master/src/parser.ts#L2000).

**Comment by 3cp** [comment]
The escaya is using old syntax OptionalExpression which doesn't require wrapping, just like the old meriyah implementation I think.

**Comment by KFlash** [comment]
Escaya are just following the ECMA specs. I updated the chaining code in escaya now to follow the specs 100%. It seems like ESTree tries to do the same, but more complex. 

Where is the wrapping?

**Comment by 3cp** [comment]
> Where is the wrapping?

https://github.com/meriyah/meriyah/blob/d3de211ef404cc10b736e853507e2f93491e4d4f/src/parser.ts#L3990-L3999

**Comment by KFlash** [comment]
https://github.com/acornjs/acorn/blob/master/acorn/src/expression.js#L300

**Comment by neumatho** [comment]
It looks like it is still needed to enable the next option for optional chaining. I think this should not be needed, since it is in the official specification now.

-Thomas

**Comment by neumatho** [comment]
There are still some issues with this. Try **a?.b[3].c?.(x).d**, then Acorn outputs this tree:

```
{
  "type": "Program",
  "start": 0,
  "end": 16,
  "body": [
    {
      "type": "ExpressionStatement",
      "start": 0,
      "end": 16,
      "expression": {
        "type": "ChainExpression",
        "start": 0,
        "end": 16,
        "expression": {
          "type": "MemberExpression",
          "start": 0,
          "end": 16,
          "object": {
            "type": "CallExpression",
            "start": 0,
            "end": 14,
            "callee": {
              "type": "MemberExpression",
              "start": 0,
              "end": 9,
              "object": {
                "type": "MemberExpression",
                "start": 0,
                "end": 7,
                "object": {
                  "type": "MemberExpression",
                  "start": 0,
                  "end": 4,
                  "object": {
                    "type": "Identifier",
                    "start": 0,
                    "end": 1,
                    "name": "a"
                  },
                  "property": {
                    "type": "Identifier",
                    "start": 3,
                    "end": 4,
                    "name": "b"
                  },
                  "computed": false,
                  "optional": true
                },
                "property": {
                  "type": "Literal",
                  "start": 5,
                  "end": 6,
                  "value": 3,
                  "raw": "3"
                },
                "computed": true,
                "optional": false
              },
              "property": {
                "type": "Identifier",
                "start": 8,
                "end": 9,
                "name": "c"
              },
              "computed": false,
              "optional": false
            },
            "arguments": [
              {
                "type": "Identifier",
                "start": 12,
                "end": 13,
                "name": "x"
              }
            ],
            "optional": true
          },
          "property": {
            "type": "Identifier",
            "start": 15,
            "end": 16,
            "name": "d"
          },
          "computed": false,
          "optional": false
        }
      }
    }
  ],
  "sourceType": "module"
}
```

Meriyah has problems with the **[3]** and **(x)** part.

Also the **({})?.a["b"]** parses wrong. It should look like this:

```
{
  "type": "Program",
  "start": 0,
  "end": 12,
  "body": [
    {
      "type": "ExpressionStatement",
      "start": 0,
      "end": 12,
      "expression": {
        "type": "ChainExpression",
        "start": 0,
        "end": 12,
        "expression": {
          "type": "MemberExpression",
          "start": 0,
          "end": 12,
          "object": {
            "type": "MemberExpression",
            "start": 0,
            "end": 7,
            "object": {
              "type": "ObjectExpression",
              "start": 1,
              "end": 3,
              "properties": []
            },
            "property": {
              "type": "Identifier",
              "start": 6,
              "end": 7,
              "name": "a"
            },
            "computed": false,
            "optional": true
          },
          "property": {
            "type": "Literal",
            "start": 8,
            "end": 11,
            "value": "b",
            "raw": "\"b\""
          },
          "computed": true,
          "optional": false
        }
      }
    }
  ],
  "sourceType": "module"
}
```

-Thomas

**Comment by 3cp** [comment]
In case I missed this, can you create a new issue or add to #110? Thx.

**Comment by 3cp** [comment]
This is fixed in latest commit in #112.
Added your test cases too.
